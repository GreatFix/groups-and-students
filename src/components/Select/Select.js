import React from 'react'
import classes from './Select.module.css'

const Select = (props) => {
  return (
    <div className={`${classes.Select} ${props.className}`}>
      {props.label && <p>{props.label}</p>}
      <select
        name={props.name}
        onChange={props.onChange}
        disabled={props.disabled}
        required={props.required}
        spellCheck={'false'}
        defaultValue={props.defaultValue}
      >
        {props.options &&
          props.options.map((item) => {
            return (
              <option key={`${item}-${Math.random()}`} value={item}>
                {item}
              </option>
            )
          })}
      </select>
    </div>
  )
}

export default Select
