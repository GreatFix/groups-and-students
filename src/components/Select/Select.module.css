.Select {
  display: flex;
  flex-direction: column;
  padding: 5px;
  margin: 4px 2px 0 2px;
}
.Select select {
  outline: none;
  border-radius: 5px;
  border: none;
  background-color: #54d8a8;
  min-height: 5vh;
  font-size: 1em;
  font-family: 'Roboto Slab', serif;
  padding: 5px;
  margin: 4px 2px 0 2px;
  cursor: pointer;
}
.Select select:focus {
  border: 1px solid #000;
  padding: 4px;
}
.Select p {
  font-size: 0.9em;
  font-family: 'Lora', serif;
  margin-left: 5px;
}
