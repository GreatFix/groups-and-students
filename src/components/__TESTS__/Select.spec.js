import { fireEvent, render } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import userEvent from '@testing-library/user-event'
import Select from '../Select/Select'

const OPTION_1 = 'option_1',
  OPTION_2 = 'option_2',
  OPTION_3 = 'option_3',
  TEST_NAME = 'TEST_NAME',
  TEST_CLASS = 'TEST_NAME',
  TEST_ID = 'TEST_ID',
  TEST_LABEL = 'TEST_LABEL'

describe('Select', () => {
  const array = [OPTION_1, OPTION_2, OPTION_3]

  test('renders with options', () => {
    const { getByText } = render(<Select name={TEST_NAME} options={array} />)
    expect(getByText(OPTION_1)).toBeInTheDocument()
    expect(getByText(OPTION_2)).toBeInTheDocument()
    expect(getByText(OPTION_3)).toBeInTheDocument()
  })

  test('gets a class', () => {
    const { container } = render(<Select name={TEST_NAME} className={TEST_CLASS} options={array} />)
    const selectParent = container.querySelector('.Select')
    expect(selectParent).toHaveClass(TEST_CLASS)
  })

  test('selects option', () => {
    const { container, getByText } = render(<Select name={TEST_NAME} options={array} />)
    const select = container.querySelector('select')
    userEvent.selectOptions(select, [OPTION_3])

    expect(getByText(OPTION_1).selected).toBe(false)
    expect(getByText(OPTION_2).selected).toBe(false)
    expect(getByText(OPTION_3).selected).toBe(true)
  })

  test('gets a id', () => {
    const { container } = render(<Select name={TEST_NAME} id={TEST_ID} />)
    const select = container.querySelector('select')
    expect(select).toHaveAttribute('id', TEST_ID)
  })

  test('gets a name', () => {
    const { container } = render(<Select name={TEST_NAME} options={array} />)
    const select = container.querySelector('select')
    expect(select).toHaveAttribute('name', TEST_NAME)
  })

  test('gets a label with id', () => {
    const { container } = render(
      <Select name={TEST_NAME} id={TEST_ID} label={TEST_LABEL} options={array} />
    )
    const label = container.querySelector('label')
    const select = container.querySelector('select')
    expect(label).toHaveTextContent(TEST_LABEL)
    expect(select).toHaveAttribute('id', TEST_ID)
  })

  test('gets a value with onChange', () => {
    const mockOnChange = jest.fn()

    const { container, getByText } = render(
      <Select name={TEST_NAME} value={OPTION_2} onChange={mockOnChange} options={array} />
    )
    const select = container.querySelector('select')

    expect(select).toHaveValue(OPTION_2)
    expect(getByText(OPTION_1).selected).toBe(false)
    expect(getByText(OPTION_2).selected).toBe(true)
    expect(getByText(OPTION_3).selected).toBe(false)
  })

  test('gets a defaultValue', () => {
    const { container, getByText } = render(
      <Select name={TEST_NAME} defaultValue={OPTION_2} options={array} />
    )
    const select = container.querySelector('select')
    expect(select).toHaveValue(OPTION_2)
    expect(getByText(OPTION_1).selected).toBe(false)
    expect(getByText(OPTION_2).selected).toBe(true)
    expect(getByText(OPTION_3).selected).toBe(false)
  })

  test('calls onChange', () => {
    const mockOnChange = jest.fn((e) => e.target.value)
    const { container, getByText } = render(
      <Select name={TEST_NAME} onChange={mockOnChange} options={array} />
    )
    const select = container.querySelector('select')

    fireEvent.change(select, {
      target: {
        value: OPTION_3,
      },
    })

    expect(mockOnChange.mock.calls.length).toBe(1)
    expect(mockOnChange.mock.results[0].value).toBe(OPTION_3)
    expect(getByText(OPTION_1).selected).toBe(false)
    expect(getByText(OPTION_2).selected).toBe(false)
    expect(getByText(OPTION_3).selected).toBe(true)
  })

  test('becomes required', () => {
    const { container } = render(<Select name={TEST_NAME} options={array} required />)
    const select = container.querySelector('select')

    expect(select).toBeRequired()
  })

  test('becomes disabled', () => {
    const { container } = render(<Select name={TEST_NAME} options={array} disabled />)
    const select = container.querySelector('select')

    expect(select).toBeDisabled()
  })

  test('render with all the props in addition to defaultValue', () => {
    const { container } = render(
      <Select
        name={TEST_NAME}
        options={array}
        id={TEST_ID}
        className={TEST_CLASS}
        label={TEST_LABEL}
        onChange={() => {}}
        value={OPTION_2}
        required
        disabled
      />
    )
    const select = container.querySelector('select')

    expect(select).toMatchSnapshot()
  })
})
