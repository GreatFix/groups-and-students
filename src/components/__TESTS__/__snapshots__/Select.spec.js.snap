// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select render with all the props in addition to defaultValue 1`] = `
<select
  disabled=""
  id="TEST_ID"
  name="TEST_NAME"
  required=""
  spellcheck="false"
>
  <option
    value="option_1"
  >
    option_1
  </option>
  <option
    value="option_2"
  >
    option_2
  </option>
  <option
    value="option_3"
  >
    option_3
  </option>
</select>
`;
